libsplinter/src/service/network/interconnect/mod.rs

```diff
@@ -13,3 +13,706 @@
+    sender: Sender<SendRequest>,
+}
+
+impl ServiceInterconnectMessageSender {
+    fn new(sender: Sender<SendRequest>) -> Self {
+        Self { sender }
+    }
+}
+
+impl MessageSender<ConnectionId> for ServiceInterconnectMessageSender {
+    fn send(
+        &self,
+        recipient: ConnectionId,
+        message: Vec<u8>,
+    ) -> Result<(), (ConnectionId, Vec<u8>)> {
+        self.sender
+            .send(SendRequest::Message(recipient, message))
+            .map_err(|err| match err.0 {
+                SendRequest::Message(recipient, payload) => (recipient, payload),
+                SendRequest::Shutdown => unreachable!(), // we didn't send this
+            })
+    }
+}
+
+impl Into<Box<dyn MessageSender<ConnectionId>>> for ServiceInterconnectMessageSender {
+    fn into(self) -> Box<dyn MessageSender<ConnectionId>> {
+        Box::new(self)
+    }
+}
+
+#[derive(Clone)]
+pub struct ShutdownHandle {
+    sender: Sender<SendRequest>,
+    dispatch_shutdown: DispatchLoopShutdownSignaler<ComponentMessageType, ConnectionId>,
+}
+
+impl ShutdownHandle {
+    /// Sends a shutdown notifications to ServiceInterconnect and the associated dipatcher thread and
+    /// Matrix
+    pub fn shutdown(&self) {
+        if self.sender.send(SendRequest::Shutdown).is_err() {
+            warn!("Service Interconnect is no longer running");
+        }
+
+        self.dispatch_shutdown.shutdown();
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+
+    use protobuf::Message;
+
+    use std::cell::RefCell;
+    use std::collections::VecDeque;
+
+    use crate::mesh::{Envelope, Mesh};
+    use crate::network::connection_manager::{
+        AuthorizationResult, Authorizer, AuthorizerError, ConnectionManager,
+    };
+    use crate::network::dispatch::{
+        ConnectionId, DispatchError, Handler, MessageContext, MessageSender,
+    };
+    use crate::protos::service;
+    use crate::service::network::ServiceConnectionManager;
+    use crate::transport::{inproc::InprocTransport, Connection, Transport};
+
+    // Verify that the ServiceInterconnect properly receives messages from services, passes them to
+    // the dispatcher, and sends messages from the handlers to other services.
+    //
+    // ServiceInterconnect will receive a message from services test-service and pass it to
+    // ComponentTestHandler. This handler will validate it came from test-service. The handler will
+    // then send a message to the ServiceInterconnect to send the message back to test-service.
+    // This valdiates that messages can be sent and recieved over the ServiceInterconnect.
+    //
+    // This tests also validates that ServiceInterconnect can retrieve the list of services from
+    // the ServiceConnectionManager using the ServiceConnectionManagerConnector.
+    //
+    // Finally, verify that ServiceInterconnect can be shutdown by calling shutdown_and_wait.
+    //
+    // 1. Starts up a ServiceConnectionManager and receives a incommint service.
```

Comment:
The steps below are not accurate for this test. Remove mentioned of NetworkEcho and "shutdown_string". Also please update the comment in the test that talk about NetworkEcho.

Reply:
Updated.